Sources/Plasma/Apps/plClient/Mac-Cocoa/plMacDisplayHelper.mm

```diff
@@ -0,0 +1,153 @@
+    // Going to be a little cheeky here - macOS has a best layout area for windows
+    // depending on dock and menu size so I'll return that instead.
+    NSRect windowArea = [screen visibleFrame];
+    fDesktopDisplayMode.Width = windowArea.size.width;
+    fDesktopDisplayMode.Height = windowArea.size.height;
+    fDesktopDisplayMode.ColorDepth = 32;
+    
+    // visibleFrame is in points - put into pixels
+    fDesktopDisplayMode.Width *= [screen backingScaleFactor];
+    fDesktopDisplayMode.Height *= [screen backingScaleFactor];
+    
+    // Calculate the region actually available for full screen
+    NSRect currentResolution = [screen frame];
+    if (@available(macOS 12.0, *)) {
+        NSEdgeInsets currentSafeAreaInsets = [screen safeAreaInsets];
+        // Sigh... Origin doesn't matter but lets do it for inspectability
+        currentResolution.origin.x += currentSafeAreaInsets.left;
+        currentResolution.origin.y += currentSafeAreaInsets.top;
+        currentResolution.size.width -= currentSafeAreaInsets.left + currentSafeAreaInsets.right;
+        currentResolution.size.height -= currentSafeAreaInsets.top + currentSafeAreaInsets.bottom;
+    }
+
+    float safeAspectRatio = currentResolution.size.width / currentResolution.size.height;
+
+    fDisplayModes.clear();
+
+    CFArrayRef displayModes = CGDisplayCopyAllDisplayModes(fCurrentDisplay, nullptr);
+    for(int i=0; i< CFArrayGetCount(displayModes); i++)
+    {
+        // Now filter out the ones that are taller than the safe area aspect ratio
+        // This could break in interesting ways if Apple ships displays that have unsafe
+        // areas along the side - but will prevent us stripping any aspect ratios that don't
+        // match the display entirely.
+        //
+        // I'd prefer not to filter a bunch of stuff out but don't strictly have a choice
+        // here because there are a lot of fake resolutions in the list.
+        //
+        // Apple also doesn't include some traditional resolutions on some of their hardware
+        // like 800x600, so don't expect those here. My Macbook Pro gives me back 1080p as the lowest.
+        // We could force that with kCGDisplayShowDuplicateLowResolutionModes but the aspect ratio
+        // filter would still drop some or all of those.
+        //
+        // Asked for better guidance here from Apple - FB13375033
+        
+        CGDisplayModeRef mode = (CGDisplayModeRef)CFArrayGetValueAtIndex(displayModes, i);
+
+        float modeAspectRatio = float(CGDisplayModeGetWidth(mode)) / float(CGDisplayModeGetHeight(mode));
+        if(modeAspectRatio < safeAspectRatio)
+        {
+            continue;
+        }
+
+        // aspect ratio is good - add the mode to the list
+        plDisplayMode plasmaMode;
+        plasmaMode.ColorDepth = 32;
+        plasmaMode.Width = int(CGDisplayModeGetWidth(mode));
+        plasmaMode.Height = int(CGDisplayModeGetHeight(mode));
```

Comment:
I know we're handling scaling elsewhere, but would it make sense for plDisplayMode to contain a scaling factor?

@Hoikas Is that something that would make any sense with Windows DPI stuff, or is that too weird?

Reply:
`plWinDpi` largely handles all of this right now, so I don't know what the value of adding a scale factor here is, but if we wanted to, `plWinDpi::GetDpi()` shows how to extract the DPI for an arbitrary `HMONITOR` on Windows 8.1, which can be converted to a scale factor by dividing by `96.f`.

```C++
    UINT dpiX, dpiY;
    HRESULT hr = GetDpiForMonitor(someHMonitorHere, MDT_EFFECTIVE_DPI, &dpiX, &dpiY);
    if (SUCCEEDED(hr)) {
        hsAssert(dpiX == dpiY, "X and Y dpi values should be the same on Windows apps");
        float scale  = (float)dpiY / 96.f; 
        return scale;
    }
    LogRed("Per-Monitor DPI failed: {}", hsCOMError(hr.value()));
    return 0.f;
```